Use `riscv` section for RISC-V targets

romancardenas opened the pull request:
This PR uses [my proposed `riscv` SVD section](https://github.com/rust-embedded/svd/pull/265) to generate RISC-V PACs using the [new set of traits in `riscv-pac`](https://github.com/rust-embedded/riscv/pull/211).

In this PR, the `mod interrupt` in RISC-V targets is handled differently, and includes:

- `ExternalInterrupt`s (very similar to current `Interrupt`s)
- `CoreInterrupt`s (this is new, previously it just assumed the peripheral follows the RISC-V standard)
- `Priority` levels (necessary for interfacing standard peripherals with trait bounds)
- `HartId`s (necessary for interfacing standard peripherals with trait bounds).

Also, the PAC does not directly generate all the handlers array and so on. Instead, it relies on macros exposed in `riscv-pac`.

I still need some time to pollish everything, but feedback is more than welcome :)

Solves #786

romancardenas commented:
[This PR](https://github.com/greenlsi/e310x/pull/1) shows how an updated PAC would look like over e310x chips

romancardenas commented:
Added the `unstable-riscv` feature to optionally activate the proposed new functionality. I still have to include support for `riscv-peripheral` implementations depending on the name of the peripheral.

Emilgardis commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/).
 ## [Unreleased]
 
+- Implement `riscv-pac` traits via experimental `riscv` element in SVD.
+  You must use the `unstable-riscv` feature to use this.
```
```suggestion
  You must compile `svd2rust` with the `unstable-riscv` feature to use this.
```

Emilgardis reviewed (commented):
good start! should probably be marked as a draft, I see a bunch of todos

romancardenas commented:
Yes! I'm currently working on a bunch of changes to use standard peripherals when possible, so it will take a few more iterations until it is ready to merge

romancardenas commented:
As shown in [this PR](https://github.com/greenlsi/e310x/pull/1), now it is possible to use `riscv-peripheral` standard implementations (i.e., CLINT and PLIC) instead of the ones generated by `svd2rust`.  [This](https://github.com/greenlsi/e310x/blob/3dcafe269c84c35c93bb2a6a5518bfc225624740/src/lib.rs#L14) shows the new outcome. Note that it only takes one macro call per standard peripheral:

```rust
riscv_peripheral :: clint_codegen ! (base 0x2000000 , mtimecmps [mtimecmp0 = (crate :: interrupt :: Hart :: H0 , "[0](crate::interrupt::Hart::H0)")] , msips [msip0 = (crate :: interrupt :: Hart :: H0 , "[0](crate::interrupt::Hart::H0)")] ,);
riscv_peripheral :: plic_codegen ! (base 0xC000000 , ctxs [ctx0 = (crate :: interrupt :: Hart :: H0 , "[0](crate::interrupt::Hart::H0)")] ,);
```

# To do list

- [ ] Configuration parameter to select whether to use the standard macros or the old-school way.
- [ ] A mechanism to specify the CLINT's clock frequency to benefit from its `embedded-hal` implementation.
- [ ] Configuration parameter to activate the use of `async_delay` flag for CLINT.

Regarding the second point: what do you think about providing the CLINT clock frequency via a configuration parameter?

romancardenas commented:
I started with the configuration parameters... and decided to explore using configuration parameters only, leaving the SVD file untouched.

For the E310x chip, the configuration file would look like [this](https://github.com/greenlsi/e310x/blob/c314fc9a53d640fad356666b364b3e7dcecfcebe/svd2rust.toml).
What do you think about this alternative? I personally like it more.

burrbull commented:
> I started with the configuration parameters... and decided to explore using configuration parameters only, leaving the SVD file untouched.
> 
> For the E310x chip, the configuration file would look like [this](https://github.com/greenlsi/e310x/blob/c314fc9a53d640fad356666b364b3e7dcecfcebe/svd2rust.toml).
> What do you think about this alternative? I personally like it more.

Makes sense. But why a TOML? YAML should looks much cleaner for such things.

Also what should we do with svd-rs changes and released crates?

burrbull commented:
I think this should not be in `svd2rust` config file. It's better to use another `--riscv-config file.yaml` option as you may want to use different configs for different input SVDs but with the same `svd2rust.toml`.

romancardenas commented:
I addressed your suggestions. I'm not sure if my independent YAML file implementation is the best approach, let me know what you think.

You can see how I use it for E310x chips [here](https://github.com/greenlsi/e310x/blob/5f39a74e5ff22d9e9d4e26aa953b40b29b1c637c/riscv_config.yaml)

romancardenas commented:
- PAC using this PR: [`e310x`](https://github.com/greenlsi/e310x/pull/1)
- HAL using the result PAC: [`e310x-hal`](https://github.com/greenlsi/e310x-hal)
- BSP using the resulting HAL: [`hifive1`](https://github.com/greenlsi/hifive1)
    - It contains an example using the PLIC and CLINT peripherals. You can emulate the CLINT example with QEMU

burrbull commented:
@romancardenas still a draft?

romancardenas commented:
Not really, I don't expect to add any changes. I'm waiting to release the next version of the `riscv` crates in crates.io, though.

burrbull commented on src/config.rs:
```diff
@@ -312,3 +314,57 @@ impl DerefMut for IdentFormats {
+#[derive(Clone, PartialEq, Eq, Debug, Default)]
+#[non_exhaustive]
+pub struct RiscvConfig {
+    pub core_interrupts: Option<Vec<RiscvEnumItem>>,
+    pub exceptions: Option<Vec<RiscvEnumItem>>,
+    pub priorities: Option<Vec<RiscvEnumItem>>,
+    pub harts: Option<Vec<RiscvEnumItem>>,
```
Is `None` different from `Some(Vec::empty)` here?

romancardenas replied:
Currently, if you define `core_interrupts` in your YAML file, `svd2rust` will generate a custom `CoreInterrupt` struct for your target, regardless of whether it is empty or not. Thus, an empty vector would probably be meaningless. I can change this behavior following any of your design decisions.

burrbull replied:
I want to say that you can leave just `Vec<T>` and check `vec.is_empty()` instead of `option.is_some`.

Possibly you should also mark a field with `serde(default)`. Not sure.

romancardenas replied:
Will take a look to your suggestion!

burrbull commented:
> I think this should not be in `svd2rust` config file. It's better to use another `--riscv-config file.yaml` option as you may want to use different configs for different input SVDs but with the same `svd2rust.toml`.

What about this?

I'd want to see something like this:
```rust
/// Common svd2rust options
struct Config {
   ...
   /// Path to YAML file with chip related settings
   pub settings: Option<PathBuf>,
}

#[non_exhaustive]
struct Settings {
    pub riscv_config: Option<RiscvConfig>,
    
    // other settings (for example #576)
}
```

romancardenas commented:
> I'd want to see something like this:
> 
> ```rust
> /// Common svd2rust options
> struct Config {
>    ...
>    /// Path to YAML file with chip related settings
>    pub settings: Option<PathBuf>,
> }
> 
> #[non_exhaustive]
> struct Settings {
>     pub riscv_config: Option<RiscvConfig>,
>     
>     // other settings (for example #576)
> }
> ```

Oh, I see! I added a `riscv_config` parameter to the general configuration, but maybe it will be more scalable having a separate `settings.yml` file that includes the RISC-V-specific configuration (together with potentially new parameters such as Zaamo-only registers).

romancardenas commented:
@burrbull Let me know if the new version looks as you expected

burrbull commented on src/generate/device.rs:
```diff
@@ -30,6 +30,14 @@ pub fn render(d: &Device, config: &Config, device_x: &mut String) -> Result<Toke
         }
     };
 
+    let _settings = match config.settings.as_ref() {
```
Why is it underscored?

burrbull commented on Cargo.toml:
```diff
@@ -35,13 +35,14 @@ default = ["bin", "json", "yaml"]
 json = ["dep:serde_json"]
 yaml = ["dep:serde_yaml"]
-unstable-riscv = ["irx-config/yaml"]
+unstable-riscv = []
```
Is it unstable yet?

burrbull commented:
Squash commits, please.

romancardenas commented:
Done

burrbull commented:
Please, revert commits from `svd-rs`.
